### ISSUE TITLE
Sensitive terms prevelance analysis in Openverse

### ISSUE THREAD

**Issue by zackkrida** [WordPress/openverse-sensitive-terms#3]
I used the following script against the production `image` index. to check the record counts of terms in the sensitive terms list. Here's the original script, where I made sure to quote terms in the same manner we do in the [index creation code](https://github.com//WordPress/openverse/blob/35100d3c719e142da2012303a5f2790c637f737a/ingestion_server/ingestion_server/indexer.py#L522).

```py
from elasticsearch import Elasticsearch
import csv

def load_terms_from_file(file_path):
    with open(file_path, 'r') as f:
        return [line.strip() for line in f]

def count_term_matches(es, index, terms, fields):
    term_counts = {}

    for term in terms:
        count = es.count(
            index=index,
            body={
                "query": {
                    "multi_match": {
                        "query": f'"{term}"',
                        "fields": fields,
                    }
                }
            }
        )['count']
        term_counts[term] = count
        print(f'Term "{term}" has {count} matches')

    return term_counts

def write_to_csv(data, file_path):
    with open(file_path, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(['Term', 'Count'])  # Writing headers
        for key, value in data.items():
            writer.writerow([key, value])

def main():
    es = Elasticsearch("http://localhost:9200")
    index = "image"
    sensitive_terms = load_terms_from_file('terms.txt')
    fields = ["tags.name", "title", "description"]

    term_counts = count_term_matches(es, index, sensitive_terms, fields)

    # Sort the dictionary by value in descending order
    sorted_term_counts = dict(sorted(term_counts.items(), key=lambda item: item[1], reverse=True))

    # Write results to CSV
    write_to_csv(sorted_term_counts, 'term_counts.csv')

if __name__ == "__main__":
    main()

```

This produced the following results: https://docs.google.com/spreadsheets/d/10w95ZZQttJplLRqxu51BEMs2pDGqtk6oq2eoGvQke7o/edit?usp=sharing

One thing I'm concerned about is that some of the highest returning results are those with multiple words and stop words. It suggests to me that some stemming may still be happening on these "exact" matches. Furthermore, one example. If I search "rosy palm and her 5 sisters" quoted in Openverse, I get no results. If I remove quotes there are many results. This search was shown to have 7,509,233 results by this script.

If I update the script to be a phrase query: 

```
 "multi_match": {
	"query": f'"{
    "type": "phrase",
    "fields": fields,
}
```

the results are very different:

https://docs.google.com/spreadsheets/d/10w95ZZQttJplLRqxu51BEMs2pDGqtk6oq2eoGvQke7o/edit#gid=945117684

This meets my expectations much more for the kind of terms which would be really prevalent. It also indicates that we should remove the term "girl on" because that's really common for innocuous things ("girl on beach" for example).

This is all I have time for, but it was really interesting and I hope it can be useful!

**Comment by sarayourfriend** [comment]
> even in quotes on openverse.org

As an aside I'm _fairly_ certain that quoting does not work via the UI. I've run into it a couple of times but forgot to dig deeper or open an issue for it.

**Comment by zackkrida** [comment]
It's so weird. 

I'm getting suspicious that our analyzer settings mean that any text query against our documents uses stemming regardless of any quoting or attempts at getting exact matches we try to do. 

For example (and apologies for the vulgar example, but for everything I tried, like "running" and "dancing", most results had both tags "dance" and "dancing" or "run" and "running"), this query:

```
{
    "query": {
        "bool": {
            "should": [
                { "term": { "tags.name.keyword": "fucking" } }
            ]
        }
    }
}

```

Returns results with that _exact_ tag. If I then do:

```
{
	"query": {
		"match_phrase": {
			"tags.name": "\"fucking\""
		}
	},
}
```

or 

```
{
	"query": {
		"match_phrase": {
			"tags.name": "\"fucking\""
		}
	},
}
```

Nearly all the results have the tag "fuck", not the tag "fucking".

**Comment by sarayourfriend** [comment]
That's interesting and concerning. I did see a significant difference when I tried the original smaller list with and without quoting though, so something should be different. But a brief bit of research reveals that it's probably a false positive. [This page talks about exact matching](https://www.elastic.co/guide/en/elasticsearch/guide/current/_finding_exact_values.html#_term_query_with_text) and indeed, it's impossible to _not_ search the tokenised (i.e., stemmed) version of `text` fields.

For tags, at least, becuase of the nature of how they usually appear (single words), we might be able to change those to check `tags.name.keyword` and avoid this issue. For the title and description, however, we might need to add new mappings for non-analysed versions. Something like `title.raw` and `description.raw` that we can send `term` against. We'd use [the `terms`](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-terms-query.html) query against those fields (and `tags.name.keyword` if we don't add a raw version of that as well), rather than multimatch, so that we get only results that exactly match the query terms. `terms` is an "OR" query, so it works for our purposes. The query would be something like:

```
{
    "query": {
        "bool": {
            "must_not": [
                { "terms": { "${field}.raw": sensitive_terms } }
                for field in ["title", "description", "tags.name"]
            ]
        }
    }
}
``` 

Interestingly, this _might_ solve the clause count issue, as it appears that term queries might use a different limit: https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#index-max-terms-count

The clause count may still come into effect though, I'll need to try it out or do a bit of research to find out.

**Comment by zackkrida** [comment]
@sarayourfriend awesome work creating the new index. The results for querying it are really compelling! Here's my script:

```
from elasticsearch import Elasticsearch
import csv

def load_terms_from_file(file_path):
    with open(file_path, 'r') as f:
        return [line.strip() for line in f]

def count_term_matches(es, index, terms, fields):
    term_counts = {}

    for term in terms:
        raw_terms_query = [{"terms": {f"{field}.raw": [term]}} for field in fields]

        count = es.count(
            index=index,
            body={
                "query": {
                    "bool": {
                        "should": raw_terms_query,
                        "minimum_should_match": 1
                    }
                }
            }
        )['count']
        term_counts[term] = count
        print(f'Term "{term}" has {count} matches')

    return term_counts

def write_to_csv(data, file_path):
    with open(file_path, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(['Term', 'Count'])  # Writing headers
        for key, value in data.items():
            writer.writerow([key, value])

def main():
    es = Elasticsearch("http://localhost:9200")
    index = "image-raw_fields"
    sensitive_terms = load_terms_from_file('terms.txt')
    fields = ["tags.name", "title", "description"]

    term_counts = count_term_matches(es, index, sensitive_terms, fields)

    # Sort the dictionary by value in descending order
    sorted_term_counts = dict(sorted(term_counts.items(), key=lambda item: item[1], reverse=True))

    # Write results to CSV
    write_to_csv(sorted_term_counts, 'term_counts_raw.csv')

if __name__ == "__main__":
    main()
```

And the results: https://docs.google.com/spreadsheets/d/10w95ZZQttJplLRqxu51BEMs2pDGqtk6oq2eoGvQke7o/edit#gid=1797307081

Here we have 1,093 terms with no match. We also have many fewer matched documents overall. And anecdotally, the terms with the highest counts seem to be much more plausible, common terms.

**Comment by sarayourfriend** [comment]
Fantastic. 840 terms still puts us over the clause limit, but I'll try a terms query and see if they're subject only to the terms limit.

**Comment by sarayourfriend** [comment]
Indeed, it seems they are separate limits. I was able to send this query:

```
{
    "query": {
        "bool": {
            "must_not": [
                { "terms": { "${field}.raw": sensitive_terms } }
                for field in ["title", "description", "tags.name"]
            ]
        }
    }
}
```

with those 840 terms and I got back results just fine. I initiated a reindex in staging to `image-raw_fields-filtered` and it didn't reject the request and appears to be running fine. Task id: `nJq9N2c-TtepEUUFsp_Yxg:616965726`.

At this point I think we have a clear change to the ingestion server that will fix the clause limits, so I'll update that issue with the solution.

At a glance, I don't even think we necessarily need to do any false positive analysis any more. Looking through the list, I agree with you, even the ones with low hits (<20) I cannot imagine being false positives.
